### PR TITLE
BAPL-1547 :Improvement the REST API in Business-Central adding the groups/roles operations available in the UI

### PR DESCRIPTION
**Pull request by RishiRajAnand**
Following new API are added to security management:
GET all groups
- http://localhost:8080/business-central/rest/groups

GET all roles
- http://localhost:8080/business-central/rest/roles

GET all users
- http://localhost:8080/business-central/rest/users

GET all EditorIds
- http://localhost:8080/business-central/rest/editors

GET all PerpectiveIds
- http://localhost:8080/business-central/rest/perpectives

GET roles by user
- http://localhost:8080/business-central/rest/users/{{username}}/roles

GET groups by user
- http://localhost:8080/business-central/rest/users/{{username}}/groups

Create group
- POST http://localhost:8080/business-central/rest/groups

Create User
- POST http://localhost:8080/business-central/rest/users

Change password User
- POST http://localhost:8080/business-central/rest/users/{username}/changePassword

DELETE User
- DELETE http://localhost:8080/business-central/rest/users/{{username}}

DELETE group
- DELETE http://localhost:8080/business-central/rest/groups/{{groupName}}

Update user roles
- POST http://localhost:8080/business-central/rest/users/{{username}}/roles

Update user groups
- POST http://localhost:8080/business-central/rest/users/{{username}}/groups

Update group permission
- POST http://localhost:8888/rest/groups/{{groupName}}/permissions

Update role permission
- POST http://localhost:8888/rest/roles/{{roleName}}/permissions

Get User permissions
- GET http://localhost:8888/rest/users/{{username}}/permissions

Get Groups permissions
- GET http://localhost:8888/rest/groups/{{groupsname}}/permissions

Get Roles permissions
- GET http://localhost:8888/rest/roles/{{rolesname}}/permissions

For detailed information about these API's refer [document](https://docs.google.com/document/d/1HLk_QNAtWSCI63A76Ou2YKLTzLKg28JsL2kYmXZVXdE/edit?usp=sharing)

Related PR's : 
https://github.com/kiegroup/kie-wb-common/pull/3315 
https://github.com/kiegroup/kie-wb-distributions/pull/1033
https://github.com/kiegroup/drools-wb/pull/1378
https://github.com/kiegroup/jbpm-wb/pull/1435
https://github.com/kiegroup/optaplanner-wb/pull/370

### CODE REVIEW

**Comment by RishiRajAnand** [comment]
Jenkins execute full downstream build

**Comment by RishiRajAnand** [comment]
Jenkins execute full downstream build

File: uberfire-api/src/main/java/org/uberfire/workbench/model/DefaultUberfireResourcesTypes.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+import org.uberfire.annotations.FallbackImplementation;
+
+/**
+ * Default implementation for {@link UberfireResourceTypes}. To override it, just provide a default
+ * CDI bean that implements {@link UberfireResourceTypes}.
+ */
+@ApplicationScoped
+@FallbackImplementation
+public class DefaultUberfireResourcesTypes implements UberfireResourceTypes {
```
[the remark this answers is not in the document]
Comment (RishiRajAnand) [reply uberfire-api/src/main/java/org/uberfire/workbench/model/DefaultUberfireResourcesTypes.java:34]: I'll change all occurances

[the remark this answers is not in the document]
Comment (RishiRajAnand) [reply uberfire-api/src/main/java/org/uberfire/workbench/model/DefaultUberfireResourcesTypes.java]: done, thanks for clarifing.

File: uberfire-api/src/main/java/org/uberfire/workbench/model/DefaultUberfireResourcesTypes.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    protected DefaultUberfireResourcesTypes() {
+    }
+
+    @Override
+    public List<String> getAllEditorId() {
```
[the remark this answers is not in the document]
Comment (RishiRajAnand) [reply uberfire-api/src/main/java/org/uberfire/workbench/model/DefaultUberfireResourcesTypes.java]: done

File: uberfire-api/src/main/java/org/uberfire/workbench/model/DefaultUberfireResourcesTypes.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (RishiRajAnand) [reply uberfire-api/src/main/java/org/uberfire/workbench/model/DefaultUberfireResourcesTypes.java]: done

File: uberfire-api/src/main/java/org/uberfire/workbench/model/DefaultUberfireResourcesTypes.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        return Arrays.asList();
+    }
+
+    @Override
+    public List<String> getAllPerpective() {
```
[the remark this answers is not in the document]
Comment (RishiRajAnand) [reply uberfire-api/src/main/java/org/uberfire/workbench/model/DefaultUberfireResourcesTypes.java]: done

**Comment by RishiRajAnand** [comment]
Jenkins retest this please

**Comment by RishiRajAnand** [comment]
Jenkins retest this please

**Comment by RishiRajAnand** [comment]
Jenkins execute full downstream build

**Comment by barboras7** [comment]
Hi @RishiRajAnand,

I tested the PR and following are my comments:

1. I believe that the methods `assignRolesToUsers()` and `assignGroupsToUsers()` in the `UserManagementResource` class should use "user" in singular as we are assigning roles/groups only to one user by this REST command
2. When creating new user by the REST API, it is not possible to also create the password for the user (if I understand correctly)
3. There is a typo in the command for getting all perspectives; now it's perpectives
`curl -u "testadmin:admin1234;" -X GET http://localhost:8080/business-central/rest/perpectives`
4. It's not possible to GET all permissions of some user/group/role (I did not notice this option anywhere) - especially compared to UI, where you can check all the permissions for given user
5. In some test, you have the "expected" value and the "actual" value in wrong order and you should swap them (so that when the tests fail in the future, it is not misleading) - you can find these places in the major issues on the SonarCloud analysis
6. It would be nice if you could cover more of the logic in tests for the `UserManagementResourceHelper` class for the public methods (for example if creating the user with null name will return a bad request), you can see this in the SonarCloud as well


Also, I found one issue in the business central, the "Edit profile preferences" is twice there and one setting is missing (I believe it's the one related with guided decision tables based on what I saw in this PR). I can create a Jira for this.
Screenshot from 2020-06-03 17-04-44


Please, note that I have not finished testing yet. I didn't have time to test the update of group permissions and role permissions but I believe I will have finished by tomorrow lunch.

**Comment by RishiRajAnand** [comment]
Thanks @barboras7 for these observations, I'll include these and push it tommorow.
As of "Edit profile preferences", it is in fact the permission for guided decision tables, its just requires changing label name. IMO we can include this as part of this too.

**Comment by barboras7** [comment]
Another two observations: 

1. `{"status":"OK","message":"Groupappraiser permissions are updated successfully."}` -> I think there's a missing space between Group and group name
2. Via REST API, I am able to request setting create=true and read=false for Pages, but it's not possible in the UI (also, when looking into the UI, I can see that this change is not done but I am getting the message about permissions updated successfully in the command line where I am running the REST API commands)

**Comment by barboras7** [comment]
Jenkins execute full downstream build

**Comment by RishiRajAnand** [comment]
@barboras7 Please download the war with new changes in this link https://drive.google.com/drive/folders/1glhqqH3gvh3a35qkT8hOuBgjQfMQpzLC
It contains all the changes mentioned by you.

**Comment by RishiRajAnand** [comment]
4 news API added
1. Change user password
2. Get User permissions
3. Get Groups permissions
4, Get Roles permissions
Please refer doc for more details

**Comment by RishiRajAnand** [comment]
Jenkins execute full downstream build

**Comment by RishiRajAnand** [comment]
Thanks @barboras7

**Comment by paulovmr** [comment]
Jenkins execute full downstream build

**Comment by paulovmr** [comment]
Unrelated test failures

**Pull request merged**